### PR TITLE
Fixed error in reverse iterators

### PR DESCRIPTION
**Pull request by Mumsfilibaba**
* Fixed error in reverse iterators that may cause a crash

### CODE REVIEW

**Comment by Mumsfilibaba** [comment]
* Also fixed a couple of errors I discovered in TSharedPtr, TWeakPtr and TUniquePtr

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 1]
```diff
@@ -311,13 +327,13 @@ namespace LambdaEngine
...
-			m_pPtr = nullptr;
-			m_pCounter = nullptr;
+			m_pPtr	= nullptr;
+			m_Counter = nullptr;
 		}
 
 	protected:
 		T* m_pPtr;
-		PtrControlBlock* m_pCounter;
+		PtrControlBlock* m_Counter;
```
Comment (TheoBerlin) [inline LambdaEngine/Include/Containers/TSharedPtr.h]: Why'd you remove the 'p' prefix?

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 2]
```diff
@@ -126,130 +130,125 @@ namespace LambdaEngine
...
+				if (m_Counter->GetStrongReferences() <= 0)
 				{
-					if (m_pCounter->GetWeakReferences() <= 0)
+					if (m_Counter->GetWeakReferences() <= 0)
 					{
-						delete m_pCounter;
+						delete m_Counter;
 					}
-
+
```
Comment (TheoBerlin) [inline LambdaEngine/Include/Containers/TSharedPtr.h:176]: :(

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 3]
```diff
@@ -126,130 +130,125 @@ namespace LambdaEngine
...
+			m_Counter	= new PtrControlBlock();
 			InternalAddStrongRef();
 		}
 
-		template<typename TOther>
+		template<typename TOther, typename DOther>
 		FORCEINLINE void InternalConstructStrong(TOther* pPtr)
 		{
-			static_assert(std::is_convertible<TOther*, T*>());
+			static_assert(std:: <TOther*, T*>());
```
Comment (TheoBerlin) [inline LambdaEngine/Include/Containers/TSharedPtr.h]:
```suggestion
			static_assert(std::<TOther*, T*>());
```

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 4]
```diff
@@ -126,130 +130,125 @@ namespace LambdaEngine
...
 
-			m_pPtr = static_cast<T*>(pPtr);
-			m_pCounter = DBG_NEW PtrControlBlock();
+			m_pPtr		= static_cast<T*>(pPtr);
+			m_Counter = new PtrControlBlock();
```
Comment (TheoBerlin) [inline LambdaEngine/Include/Containers/TSharedPtr.h]: Why not use `DBG_NEW`?

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 5]
```diff
@@ -418,14 +448,15 @@ namespace LambdaEngine
 
 		FORCEINLINE T* operator->() const noexcept
 		{
-			return Get();
+			return TBase::Get();
 		}
-
+
```
Comment (TheoBerlin) [inline LambdaEngine/Include/Containers/TSharedPtr.h:455]: Trailing boys at it again.

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 6]
```diff
@@ -603,12 +648,12 @@ namespace LambdaEngine
...
-		FORCEINLINE T& operator[](uint32 Index) const noexcept
+		FORCEINLINE T& operator[](uint32 index) noexcept
 		{
-			VALIDATE(m_pPtr != nullptr);
-			return m_pPtr[Index];
+			VALIDATE(TBase::m_pPtr != nullptr);
+			return TBase::m_pPtr[index];
 		}
-
+
```
Comment (TheoBerlin) [inline LambdaEngine/Include/Containers/TSharedPtr.h:658]: Trailing boy

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 7]
```diff
@@ -1057,21 +1105,98 @@ namespace LambdaEngine
...
+		return ::Move(TSharedPtr<T>(pRefCountedPtr));
+	}
+
+	/*
+	* Casting functions
+	*/
+
+	// static_cast
+	template<typename T0, typename T1>
+	std::enable_if_t<std::is_array_v<T0> == std::is_array_v<T0>, TSharedPtr<T0>> StaticCast(const TSharedPtr<T1>& pPointer)
```
Comment (skumbananaaa) [inline LambdaEngine/Include/Containers/TSharedPtr.h]: T0 == T1

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 8]
```diff
@@ -1057,21 +1105,98 @@ namespace LambdaEngine
...
+	{
+		using TType = TRemoveExtent<T0>;
+		
+		TType* pRawPointer = static_cast<TType*>(pPointer.Get());
+		return ::Move(TSharedPtr<T0>(pPointer, pRawPointer));
+	}
+
+	template<typename T0, typename T1>
+	std::enable_if_t<std::is_array_v<T0> == std::is_array_v<T0>, TSharedPtr<T0>> StaticCast(TSharedPtr<T1>&& pPointer)
```
Comment (skumbananaaa) [inline LambdaEngine/Include/Containers/TSharedPtr.h]: T0 == T1

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 9]
```diff
@@ -1057,21 +1105,98 @@ namespace LambdaEngine
...
+	{
+		using TType = TRemoveExtent<T0>;
+		
+		TType* pRawPointer = static_cast<TType*>(pPointer.Get());
+		return ::Move(TSharedPtr<T0>(::Move(pPointer), pRawPointer));
+	}
+
+	// const_cast
+	template<typename T0, typename T1>
+	std::enable_if_t<std::is_array_v<T0> == std::is_array_v<T0>, TSharedPtr<T0>> ConstCast(const TSharedPtr<T1>& pPointer)
```
Comment (skumbananaaa) [inline LambdaEngine/Include/Containers/TSharedPtr.h]: T0 == T1

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 10]
```diff
@@ -1057,21 +1105,98 @@ namespace LambdaEngine
...
+	{
+		using TType = TRemoveExtent<T0>;
+		
+		TType* pRawPointer = const_cast<TType*>(pPointer.Get());
+		return ::Move(TSharedPtr<T0>(pPointer, pRawPointer));
+	}
+
+	template<typename T0, typename T1>
+	std::enable_if_t<std::is_array_v<T0> == std::is_array_v<T0>, TSharedPtr<T0>> ConstCast(TSharedPtr<T1>&& pPointer)
```
Comment (skumbananaaa) [inline LambdaEngine/Include/Containers/TSharedPtr.h]: T0 == T1

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 11]
```diff
@@ -1057,21 +1105,98 @@ namespace LambdaEngine
...
+	{
+		using TType = TRemoveExtent<T0>;
+		
+		TType* pRawPointer = const_cast<TType*>(pPointer.Get());
+		return ::Move(TSharedPtr<T0>(::Move(pPointer), pRawPointer));
+	}
+
+	// reinterpret_cast
+	template<typename T0, typename T1>
+	std::enable_if_t<std::is_array_v<T0> == std::is_array_v<T0>, TSharedPtr<T0>> ReinterpretCast(const TSharedPtr<T1>& pPointer)
```
Comment (skumbananaaa) [inline LambdaEngine/Include/Containers/TSharedPtr.h]: T0 == T1

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 12]
```diff
@@ -1057,21 +1105,98 @@ namespace LambdaEngine
...
+	{
+		using TType = TRemoveExtent<T0>;
+		
+		TType* pRawPointer = reinterpret_cast<TType*>(pPointer.Get());
+		return ::Move(TSharedPtr<T0>(pPointer, pRawPointer));
+	}
+
+	template<typename T0, typename T1>
+	std::enable_if_t<std::is_array_v<T0> == std::is_array_v<T0>, TSharedPtr<T0>> ReinterpretCast(TSharedPtr<T1>&& pPointer)
```
Comment (skumbananaaa) [inline LambdaEngine/Include/Containers/TSharedPtr.h]: T0 == T1

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 13]
```diff
@@ -1057,21 +1105,98 @@ namespace LambdaEngine
...
+	{
+		using TType = TRemoveExtent<T0>;
+		
+		TType* pRawPointer = reinterpret_cast<TType*>(pPointer.Get());
+		return ::Move(TSharedPtr<T0>(::Move(pPointer), pRawPointer));
+	}
+
+	// dynamic_cast
+	template<typename T0, typename T1>
+	std::enable_if_t<std::is_array_v<T0> == std::is_array_v<T0>, TSharedPtr<T0>> DynamicCast(const TSharedPtr<T1>& pPointer)
```
Comment (skumbananaaa) [inline LambdaEngine/Include/Containers/TSharedPtr.h]: T0 == T1

File: LambdaEngine/Include/Containers/TSharedPtr.h [hunk 14]
```diff
@@ -1057,21 +1105,98 @@ namespace LambdaEngine
...
+	{
+		using TType = TRemoveExtent<T0>;
+		
+		TType* pRawPointer = dynamic_cast<TType*>(pPointer.Get());
+		return ::Move(TSharedPtr<T0>(pPointer, pRawPointer));
+	}
+
+	template<typename T0, typename T1>
+	std::enable_if_t<std::is_array_v<T0> == std::is_array_v<T0>, TSharedPtr<T0>> DynamicCast(TSharedPtr<T1>&& pPointer)
```
Comment (skumbananaaa) [inline LambdaEngine/Include/Containers/TSharedPtr.h]: T0 == T1

**Review by TheoBerlin — APPROVED** [review]

**Pull request merged**